packages/micropip/src/micropip/_micropip.py

```diff
@@ -86,23 +90,28 @@ def is_compatible(self):
                 return True
         return False
 
-    async def download(self, fetch_kwargs):
+    async def _fetch_bytes(self, fetch_kwargs):
         try:
-            wheel_bytes = await fetch_bytes(self.url, fetch_kwargs)
+            return await fetch_bytes(self.url, fetch_kwargs)
         except OSError as e:
-            if self.url.startswith("https://files.pythonhosted.org/"):
+            if self.parsed_url.hostname == "files.pythonhosted.org":
                 raise e
             else:
                 raise ValueError(
-                    f"Can't fetch wheel from '{self.url}'."
+                    f"Can't fetch wheel from '{self.url}'. "
                     "One common reason for this is when the server blocks "
-                    "Cross-Origin Resource Sharing (CORS)."
+                    "Cross-Origin Resource Sharing (CORS). "
                     "Check if the server is sending the correct 'Access-Control-Allow-Origin' header."
                 ) from e
 
-        self.data = BytesIO(wheel_bytes)
-
-        with ZipFile(self.data) as zip_file:
+    async def download(self, fetch_kwargs):
+        data: BytesIO | BufferedReader
+        if self.parsed_url.scheme == "emfs":
+            data = open(self.parsed_url.path, "rb")
```

Comment:
I wonder if it wouldn't be better to directly add this to fetch_bytes (used in `_fetch_bytes`). Yes, it's potentially one extra copy but that function already handles `file://` protocol, so it would make sense to have a function that is able to open URLs with different protocols rather than have this logic through the code.

No strong opinion about this though.

Reply:
This is even possible without an extra copy if we make a few improvements to the code.